401 login page and 403 page

Atmire-Kristof opened the pull request:
## References
Fixes #923 

## Description
This PR changes the `redirectOn404Or401` (now called `redirectOn4xx`) operator to redirect the user to the login page when receiving a 401 and the user isn't logged in. It now also supports a redirect to a 403 page.

## Instructions for Reviewers
How to test:
- Visit a restricted dso while logged out. Visiting the page should redirect you to the login page and after logging in, you should end up on the requested page.
- Visit a forbidden dso while logged in as a user that doesn't have any rights to view this dso. You should be redirected to a 403 page.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

tdonohue commented on src/app/core/shared/operators.ts:
```diff
@@ -178,18 +179,29 @@ export const getAllSucceededRemoteListPayload = () =>
+              if (rd.error.statusCode === 403) {
+                router.navigateByUrl(getForbiddenRoute(), {skipLocationChange: true});
+              } else if (rd.error.statusCode === 401) {
+                router.navigateByUrl(getUnauthorizedRoute(), {skipLocationChange: true});
```
Silly question perhaps, but how can a 401 ever occur when you are already authenticated (i.e. `if (isAuthenticated)`)?  If that ever occurs, that'd imply a major flaw in the REST API -- as a 401 means you are *not authenticated* or failed to authenticate. 

Do we even need `getUnauthorizedRoute()` anymore, or is this just a safety measure of some sort?  It also makes me wonder if we need the `UnauthorizedComponent` anymore, as it seems like it should never be used.

artlowel replied:
You're right, we'll remove it

tdonohue commented on src/app/app-routing.module.ts:
```diff
@@ -14,14 +14,15 @@ import {
   ADMIN_MODULE_PATH,
   BITSTREAM_MODULE_PATH,
-  INFO_MODULE_PATH
+  INFO_MODULE_PATH, FORBIDDEN_PATH
```
Tiny thing...just for readability, it might be nice to put this new constant on it's own line.

artlowel replied:
In general we don't look at the imports. Our IDE manages them automatically and even keeps them hidden by default these days. I understand a reviewer asking to remove unused imports, or not to reorder unnecessarily as that makes PRs seem bigger than they actually are. 

However manually going over the imports the IDE has added to determine whether an import should go on a new line or not doesn't seem like a very good use of a developer's time.

tdonohue replied:
@artlowel : makes sense. I wasn't sure if this was something that is an IDE configuration that could be tweaked.  But, no big deal to let these imports align however the IDE decides to do them.  So, I agree with you, this is not really worth the time, so I'll refrain from these comments in the future ;)

tdonohue reviewed (commented):
@Atmire-Kristof : Overall this looks great & it works as described.  That said, I noticed a small oddity in the code of `operators.ts` (see inline comment below) which I'd like to get your feedback & @artlowel 's feedback on.  It looks (to me) like we may no longer need the `UnauthorizedComponent` (at all), as it's only used in one place which seems impossible to trigger.  Maybe I'm missing something here though.

In any case, I'm generally +1 on this PR.  I'd just like you to look closer at the code to see if we should clean this up further and remove the `UnauthorizedComponent`.

Atmire-Kristof commented:
As @tdonohue mentioned, it doesn't make sense for the operator to be able to redirect to a 401 page if not being logged in would redirect you to the login page regardless, so I've removed this check in the operator.
I also realised the feature-authorization guards use the 401 page when receiving a 401 from the REST API. I've modified this part to redirect to the login page instead, for more consistency. This means the 401 page is now completely useless, so I've also removed it and every use of it entirely.

tdonohue reviewed (approved):
👍 Code looks good now, and I re-tested and it works as expected.  Thanks @Atmire-Kristof !
